### ISSUE TITLE
'dynamic' keyword not working as expected

### ISSUE THREAD

**Issue by JLEYDET** [dotnet/runtime#107512]
In .NET 4.8, this works...
```
var _session = (ExpectedType)Marshal.GetActiveObject("Some.Application");
Console.WriteLine(_session.SomeProperty.FullName);
```
... but this throws an exception:
```
dynamic _session = Marshal.GetActiveObject("Some.Application");
Console.WriteLine(_session.SomeProperty.FullName);
```

I have the exact same problem in .NET 8. (The only difference is that I'm porting Marshal.GetActiveObject which is missing in .NET 8.)

I'm not sure if that's an issue with the .NET runtime, if I'm simply misusing the _dynamic_ keyword, or if the automation server is the culprit.

According to the application's docs all COM interfaces exposed by that application should implement _IDispatch_. In fact, I can use _IDispatch.GetTypeInfo_ and _ITypeInfo.GetDocumentation_ to retrieve the type name of the objects, and _Type.InvokeMember_ to interact with the automation server in a late-bound style.

The application vendor doesn't formally support automation from .NET, but I appreciate any suggestion for what to try next, that will help narrow down the problem (and possibly start a discussion with the vendor if it turns out their product is broken - this wouldn't be the first time).

Many thanks.

**Comment by colejohnson66** [comment]
What error do you get when using `dynamic`?

**Comment by JLEYDET** [comment]
> What error do you get when using `dynamic`?

```
System.Runtime.InteropServices.COMException
  HResult=0x80004005
  Message=Error HRESULT E_FAIL has been returned from a call to a COM component.
  Source=mscorlib
  StackTrace:
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHRInternal(Int32 errorCode, IntPtr errorInfo)
   at System.Dynamic.ComRuntimeHelpers.GetITypeInfoFromIDispatch(IDispatch dispatch, Boolean throwIfMissingExpectedTypeInfo)
   at System.Dynamic.IDispatchComObject.EnsureScanDefinedMethods()
   at System.Dynamic.IDispatchComObject.System.Dynamic.IDynamicMetaObjectProvider.GetMetaObject(Expression parameter)
   at System.Dynamic.DynamicMetaObject.Create(Object value, Expression expression)
   at System.Dynamic.DynamicMetaObjectBinder.Bind(Object[] args, ReadOnlyCollection`1 parameters, LabelTarget returnLabel)
   at System.Runtime.CompilerServices.CallSiteBinder.BindCore[T](CallSite`1 site, Object[] args)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
   at dynamic_keyword.Program.Main(String[] args) in C:\...\Program.cs:line 17

```

**Comment by colejohnson66** [comment]
I'm not well versed in COM, but it looks like your library type implements `IDispatch`, but when the runtime asks it what its type is (to look up the vtable?), `IDispatch::GetTypeInfoCount` returns `E_FAIL`. Which would explain why casting it first works; The runtime then knows what its type is.

**Comment by JLEYDET** [comment]
Thank you for your suggestions.
`IDispatch` was implemented indeed, and _seemingly_ behaving correctly (I could reflect on the `System.__ComObject` Type with `Type.InvokeMember`).

This morning I remembered facing a minor issue with Visual Studio some time ago, where the debugger couldn't inspect COM objects for the current release of the application (which is the release I'm using right now). So I got a hold on a newer version of the application where I know the COM debugging issues were fixed and guess what... The issue being discussed here is gone in that version too!

I don't know COM well enough to understand how an `IDispatch` implementation could mostly work but not quite, but I guess it's good to know that it can be the case.
